### ISSUE TITLE
Switch to self-hosted GitHub Actions for CI

### ISSUE THREAD

**Issue by kring** [CesiumGS/cesium-unreal#624]
Our current Travis CI system works pretty well for packaging the plugin, but it has some pain points:
* We can't customize the image used to run our builds, so the build process has to copy Unreal Engine, the Android NDK, and whatever else it needs on every build. This takes a significant amount of time.
* Travis VMs have limited disk space. Unreal Engine _barely_ fits, and even then only by deleting a bunch of stuff from it, which is super error prone and tedious.
* It takes forever to build a new Unreal Engine distribution for use on Travis, which we need to do semi-often for new UE versions and new platforms.
* We're able to package plugins, but packaging entire applications fails at the cooking step. We're not too sure why, but it's likely either a limitation of the Travis VM or missing pieces in our hacked up UE distribution.

I spent some time looking at switching to GitHub Actions. Using the standard GitHub-hosted runners is a non-starter, because the VMs have even less available disk space than the Travis VMs. No way we can squeeze Unreal Engine in there. But self-hosted runners should work really well. Here's how I think we can approach it:

1. Create a [Packer](https://www.packer.io/) script to start with the latest Windows AMI, install Unreal Engine (well, probably download and extract a ZIP rather than install, because I don't think official UE builds can be installed in an automated fashion), and create a new AMI from it.
2. Use https://github.com/machulav/ec2-github-runner in our GitHub Actions workflow to automatically start a brand new EC2 instance with the above AMI, run our build on it, and then shut it down.

We want to automate creation of the AMI - rather than just doing it once manually - because we need to repeat the process often (weekly?) in order to get new OS security patches. We may be able drive this part with GitHub Actions too (there are Packer GitHub Actions).

The nice part of this approach is that it can scale to run any number of builds simultaneously (if we want), but we only pay for EC2 time while builds are running. It looks like GitHub charges $0.016 per minute for their hosted runners, with up to 3000 free minutes (worth $48) included in some of the GitHub plans. We can use some very beefy EC2 instances for that price, so we should be able to either save money or go faster (or both).

Because this is a public repo, we'll need to be very careful, because pull requests can run code on our EC2 instances. Because we're starting with a clean EC2 and throwing it away each time, there's no risk of the machine itself being compromised beyond its lifetime. But we need to make sure that the EC2 doesn't have any privileged AWS or network access. And even so, a malicous job could mine bitcoin or send spam or whatever. Perhaps we can disable the workflow (or use an alternate one) for PRs from outside the main repo.

I considered other options, too, like using [ue4-docker](https://github.com/adamrehn/ue4-docker) plus ECS / Fargate, but that setup seems less flexible and more difficult to set up. For example, see https://dev.to/jimmydqv/github-self-hosted-runners-in-aws-part-1-fargate-39hi.

**Comment by xuelongmu** [comment]
Looks like Windows 2019 environments in Github Actions have been upgraded so there should be plenty of space for UE4 + dependencies: https://github.com/actions/virtual-environments/pull/309

**Comment by xuelongmu** [comment]
Also Github Actions is free on public repos: https://docs.github.com/en/actions/learn-github-actions/usage-limits-billing-and-administration

**Comment by mramato** [comment]
> But because the windows server image + all prerequisites is getting so big, they had to upsize from a 128gb VM to 256gb, so there should be at least 128gb of free space.

This sounds like a risky gamble to make.  If they suddenly start enforcing the limit, we would be screwed,

**Comment by xuelongmu** [comment]
@kring I checked the disk space in a github actions Windows 2019 VM and can confirm there's a good amount of disk space:
```
2021-12-02T01:06:44.9390000Z DriveLetter FriendlyName      FileSystemType DriveType HealthStatus OperationalStatus SizeRemaining      Size
----------- ------------      -------------- --------- ------------ ----------------- -------------      ----
D           Temporary Storage NTFS           Fixed     Healthy      OK                     12.19 GB     14 GB
A                             Unknown        Removable Healthy      Unknown                     0 B       0 B
            System Reserved   NTFS           Fixed     Healthy      OK                    465.45 MB    500 MB
C           Windows           NTFS           Fixed     Healthy      OK                     88.27 GB 255.51 GB
```
SizeRemaining is 88.27 GB

**Comment by xuelongmu** [comment]
Could you describe the process for creating that UE distribution that gets uploaded to Google Cloud? I'd like to replicate it and create a version that has no missing pieces because there's more space in the Github Actions VM.

**Comment by kring** [comment]
@xuelongmu there's a readme-cesium.txt in the ZIP that documents the process. I'll paste it here to save you the trouble of downloading the giant file:

```
This distribution was created by switching the Epic Launcher to PCB Mode (https://docs.unrealengine.com/4.27/en-US/Basics/InstallingUnrealEngine/MultipleLauncherInstalls/) and then installing the standard Unreal Engine v4.27 via the Marketplace.

After installation, we deleted the following to reduce the size:
- The Engine/Content directory
- Everything in Engine/Plugins/ except for the following:
  - Engine/Plugins/Editor/BlueprintMaterialTextureNodes
  - Engine/Plugins/Editor/ContentBrowser
  - Engine/Plugins/Experimental/GeometryProcessing
  - Engine/Plugins/Experimental/Landmass
  - Engine/Plugins/Experimental/PythonScriptPlugin
  - Engine/Plugins/Experimental/Water
  - Engine/Plugins/FX/Niagara
  - Engine/Plugins/Messaging/TcpMessaging
  - Engine/Plugins/Runtime/AndroidPermission
  - Engine/Plugins/Runtime/SunPosition
  - Engine/Plugins/Runtime/WebBrowserWidget
- *.pdb in all directories

Then we ZIPed it up using 7zip with Ultra compression.
```

That's for UE 4.27. 4.26 should be much the same. The set of plugins may be slightly different, but for your purposes you probably just want to keep them all anyway.

**Comment by kring** [comment]
This was implemented for UE5 in #811.
